### PR TITLE
Fix [Bug] Splitting Dataset when shards < n hangs or errors

### LINKED ISSUE #19644: [Bug] Splitting DatasetPipeline when shards < n hangs or errors
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Others

### What happened + What you expected to happen

I have a small test dataset I'm running against. Because the number of rows is very small for testing (25), it can lead to scenarios where the dataset in the pipeline cannot be split into 2 shards. If I set `equal=False` (the default), then the program will hang with the following message:

```
(BaseWorkerMixin pid=79809) Warning: reader on shard 1 of the pipeline has been blocked more than 10.0s waiting for other readers to catch up. All pipeline shards must be read from concurrently.
(BaseWorkerMixin pid=79809) Warning: reader on shard 1 of the pipeline has been blocked more than 20.0s waiting for other readers to catch up. All pipeline shards must be read from concurrently.
(BaseWorkerMixin pid=79809) Warning: reader on shard 1 of the pipeline has been blocked more than 40.0s waiting for other readers to catch up. All pipeline shards must be read from concurrently.
(BaseWorkerMixin pid=79809) Warning: reader on shard 1 of the pipeline has been blocked more than 80.0s waiting for other readers to catch up. All pipeline shards must be read from concurrently.
```

If, however, I set `equal=True` I get this error:

```
Traceback (most recent call last):
  File "/Users/tgaddair/repos/ludwig/tests/integration_tests/utils.py", line 435, in _subproc_wrapper
    results = fn(*args, **kwargs)
  File "/Users/tgaddair/repos/ludwig/tests/integration_tests/test_ray.py", line 101, in run_test_parquet
    run_fn(config, data_parquet=dataset_parquet)
  File "/Users/tgaddair/repos/ludwig/tests/integration_tests/test_ray.py", line 72, in run_api_experiment
    train_with_backend(dask_backend, config, dataset=data_parquet, evaluate=False)
  File "/Users/tgaddair/repos/ludwig/tests/integration_tests/utils.py", line 651, in train_with_backend
    skip_save_log=True,
  File "/Users/tgaddair/repos/ludwig/ludwig/api.py", line 538, in train
    save_path=model_dir,
  File "/Users/tgaddair/repos/ludwig/ludwig/backend/ray.py", line 250, in train
    **kwargs
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/trainer.py", line 240, in run
    for intermediate_result in iterator:
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/trainer.py", line 568, in __next__
    self._executor.finish_training)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/trainer.py", line 537, in _run_with_error_handling
    return func()
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/backends/backend.py", line 600, in finish_training
    results = self.get_with_failure_handling(futures)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/backends/backend.py", line 619, in get_with_failure_handling
    success, failed_worker_indexes = check_for_failure(remote_values)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/utils.py", line 34, in check_for_failure
    ray.get(object_ref)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 89, in wrapper
    return func(*args, **kwargs)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/worker.py", line 1621, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AssertionError): ray::BaseWorkerMixin._BaseWorkerMixin__execute() (pid=79959, ip=192.168.4.54, repr=<ray.util.sgd.v2.worker_group.BaseWorkerMixin object at 0x15093c250>)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/worker_group.py", line 25, in __execute
    return func(*args, **kwargs)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/backends/backend.py", line 574, in end_training
    output = session.finish()
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/session.py", line 84, in finish
    func_output = self.training_thread.join()
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/utils.py", line 86, in join
    raise self.exc
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/utils.py", line 79, in run
    self.ret = self._target(*self._args, **self._kwargs)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/util/sgd/v2/trainer.py", line 337, in <lambda>
    return lambda: train_func(config)
  File "/Users/tgaddair/repos/ludwig/ludwig/backend/ray.py", line 246, in <lambda>
    lambda config: train_fn(**config),
  File "/Users/tgaddair/repos/ludwig/ludwig/backend/ray.py", line 197, in train_fn
    training_set_metadata,
  File "/Users/tgaddair/repos/ludwig/ludwig/data/dataset/ray.py", line 127, in __init__
    self.dataset_iter = dataset_shard.iter_datasets()
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/data/dataset_pipeline.py", line 300, in iter_datasets
    return PipelineExecutor(self)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/data/impl/pipeline_executor.py", line 28, in __init__
    self._stages[0] = pipeline_stage.remote(next(self._iter))
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/data/dataset_pipeline.py", line 220, in __next__
    self.split_index))
ray.exceptions.RayTaskError(AssertionError): ray::PipelineSplitExecutorCoordinator.next_dataset_if_ready() (pid=79974, ip=192.168.4.54, repr=<ray.data.impl.pipeline_executor.PipelineSplitExecutorCoordinator object at 0x117a5e510>)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/data/impl/pipeline_executor.py", line 100, in next_dataset_if_ready
    self.cur_splits = self.splitter(ds)
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/data/dataset_pipeline.py", line 148, in <lambda>
    lambda ds: ds.split(n, equal=equal, locality_hints=locality_hints))
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/data/dataset.py", line 531, in split
    for actor in locality_hints
  File "/Users/tgaddair/.venv/ludwig/stable/lib/python3.7/site-packages/ray/data/dataset.py", line 418, in equalize
    assert lower_bound > 0, splits
AssertionError: [Dataset(num_blocks=1, num_rows=1, schema={audio_E7982_3_xuCI: list<item: float>, binary_4523E_mZFLky: bool, __index_level_0__: int64}), Dataset(num_blocks=1, num_rows=0, schema={audio_E7982_3_xuCI: null, binary_4523E_mZFLky: bool, __index_level_0__: int64})]
```

I think both situations could use a better error message, though in the first case, I would expect it should work. If there are two workers and only one shard worth of data to be split, one of the workers should just get an empty dataset, and processing should continue normally.

### Versions / Dependencies

ray==1.7.0

### Reproduction script

Split on a DatasetPipeline that has only one shard / partition and distribute it to two workers.

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Some point fixes to split() to resolve the edge cases. Moves the test cases for split() into a new file, and adds a sweep across the corner cases in test_split_small.

Closes https://github.com/ray-project/ray/issues/19644

### CODE REVIEW

File: python/ray/data/tests/test_split.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+                if sum(block_sizes) < num_splits:
+                    min_ = math.ceil(num_splits / num_blocks)
+                    block_sizes = np.random.randint(
+                        min_, max_num_rows_per_block + 1, size=num_blocks
+                    )
+                _test_equal_split_balanced(block_sizes, num_splits)
+
+
+@pytest.mark.parametrize("pipelined", [False, True])
+def test_split_small(ray_start_regular_shared, pipelined):
```
Comment (ericl) [inline python/ray/data/tests/test_split.py:141]: This is the new test case.

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -622,7 +626,7 @@ def _equalize_larger_splits(
             leftovers = []
             for split in splits:
                 size = counts_cache[split._get_uuid()]
-                if size == target_size:
+                if size == target_size or target_size == 0:
```
Comment (ericl) [inline python/ray/data/dataset.py:629]: This isn't exactly a proper fix, but the simplest one for now. I am thinking of rewriting split() to clean up the logic.

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -622,7 +626,7 @@ def _equalize_larger_splits(
...
                     new_splits.append(split)
                     continue
                 split_indices = list(range(target_size, size, target_size))
```
Comment (ericl) [inline python/ray/data/dataset.py:632]: Previously we'd crash here if target_size == 0.

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -759,10 +763,11 @@ def equalize(splits: List[Dataset[T]], num_splits: int) -> List[Dataset[T]]:
                         self._lazy,
                     )
                     for blocks in np.array_split(block_refs, n)
-                    if not equal or len(blocks) > 0
```
Comment (ericl) [inline python/ray/data/dataset.py:762]: Remove this since we should always return the same number of outputs.

**Comment by jjyao** [comment]
Could you change the title to `Splitting Dataset when shards < n hangs or errors`, I think it has nothing to do with DatasetPipeline?

**Review by clarkzinzow — APPROVED** [review]
LGTM
